### PR TITLE
ENH: Open Document Format ODS support in read_excel()

### PR DESCRIPTION
**Pull request by davidovitch**
This work should eventually lead to a reader and writer for ods spreadsheets (LibreOffice / OpenOffice Open Document Format). I am doing this work in the master of my clone of pandas. From the wiki I understood it is fine for me to commit to my master and rebase with upstream from time to time. Please let me know if I should follow another git workflow (this is my first PR).

This is what I have so far:
- It seems to make sense to implement an OdsFile and OdsWriter in pandas/io/excel.py
- I created ods test files identical to the other Excel test files.
- Basic reading works, but does not handle all corner cases correctly yet
- Tests and WritingOds are not implemented yet

I am not sure how to approach the following issues:
- For the tests, it would make sense to also use pandas/io/tests/test_excel.py (there would be a lot of code duplication otherwise). However, this would require some kind of a more elaborate test skipping mechanism than currently implemented. One could have ods dependencies installed, but not the Excel ones and vice versa. 
- When considering pandas could support more spreadsheet formats, does it make sense to rename pandas/io/excel.py to spreadsheet.py, and correspondingly, read_excel to read_spreadsheet? I guess this is not easy since it a lot of stuff out there already counts on the excel naming scheme.

For the record, there is one message on the [mailing list](https://groups.google.com/forum/?fromgroups#!topic/pydata/Zv-BXJNsP2U), and this PR will fix issue #2311.

ods support depends on [ezodf](https://github.com/T0ha/ezodf), and issue #2311 mentions some of the alternatives.

### CODE REVIEW

**Comment by davidovitch** [comment]
So, after a long pause I am trying to get this going again. I just did a rebase with upstream and pushed my latest changes to my remote clone here. However, I think I did something not entirely correct because now I can see some other commits appearing in this pull request... 

@jreback, i have made an attempt to make it less messy, but I have to admit I wasn't completely sure I understood your suggestions correctly. I guess you'll let me know if this is not exactly right :-)

There is now another test failing, so this is not over yet. Anyway, thanks for reviewing this. It has been an interesting and learnfull ride so far.

**Comment by jreback** [comment]
@davidovitch ok, first order of biz is to rebase this. Then can see what the changes.

**Comment by davidovitch** [comment]
@jreback, I think git history is fixed now. All but the most recent commit were doubled previously. I have removed those commits with a rebase, followed by a push origin master -f to here.

File: pandas/io/excel.py [hunk 1]
```diff
@@ -66,17 +69,88 @@ def get_writer(engine_name):
...
+            _readers[self.engine] = False
+
+
+class XLRDFile(BaseFile):
+
+    def __init__(self, **kwargs):
+        self.engine = 'xlrd'
+        self.extensions = ['xls', 'xlsx', 'xlsm']
+        self.io_class = type(None)
+        self.open_workbook = None
```
Comment (jreback) [inline pandas/io/excel.py]: I would pass all of these arguments to the ``BaseFile.__init__`` and have it do the assignments

File: pandas/io/excel.py [hunk 2]
```diff
@@ -66,17 +69,88 @@ def get_writer(engine_name):
...
+        super(XLRDFile, self).__init__(**kwargs)
+
+    def load_engine(self):
```
Comment (jreback) [inline pandas/io/excel.py]: maybe define ``load_engine`` in ``BaseFile`` and have it raise ``NotImplementedError(...)`` to signal the sub-classes should impement

**Comment by davidovitch** [comment]
@jreback, is this looking better? Regarding the tests for the readers (ExcelReaderTests), I added a method to determine which readers are installed, and only test for those. This is different then before where a test would skip if only one out of 3 readers was not installed. I think it looks less compact in the code, but should result in better test coverage when not all the spreadsheet readers are installed.

**Comment by davidovitch** [comment]
@shoyer, @jreback I would like to bring this PR back to life. Is any one of you interested in another round of reviews?

**Comment by shoyer** [comment]
Sure, let us know when you're ready

**Comment by davidovitch** [comment]
@shoyer, @jreback: I've cleaned it up a bit more. Personally, I believe this PR has come along way, but it is far from being perfect:
* No writer capabilities yet, and I think that should be another PR
* Not all file type determination logic is out of the ```ExcelFile.__init__()``` class. I think it would be more elegant if BaseFile and its subclasses (```EZODFFile``` and ```XLRDFile```) should contain the file format/library dependent logic.
* ```ExcelFIle``` still contains ```ezodf``` and ```xlrd``` dependent methods: ```_parse_ods```, ```_parse_datetime()```, and ```_print_ods_cellinfo()``` for ```ezodf``` and ```_parse_excel()``` for ```xlrd```. I would assume it to be more elegant if those somehow could live in the respective ```BaseFile``` subclasses ```EZODFile``` and ```XLRDFile```
* A few days ago I bumped into [pyexcel](https://github.com/chfw/pyexcel): a library with a single API to read all the different spreadsheet formats. It depends on the same libraries pandas is using for reading/writing of the different formats, and I am wondering if the work I am doing here makes any sense if, instead of implementing this for all the different readers/writers out there, we could focus on supporting one library that handles all the format dependent logic/interpretation for us. I also mention this in issue #2311. Any thoughts on this?

**Comment by jreback** [comment]
@davidovitch seems that ``pyexcel`` uses lots of different libraries (well plugins), and thus to add ``ods`` support you would need: ``pyexcel, pyexcel-io, texttable, pyexcel-ods, pyexcel-ods3`` is that accurate? if that is the case then too many deps.

**Comment by davidovitch** [comment]
@jreback you are right: ```pyexcel ``` uses a plugin infrastructure. The plugins itself are very small (the actual ```pyexcel-ods``` code doing stuff is only 250 lines approximately). Since ```pyexcel``` uses a new BSD license (I am not a lawyer, but new BSD and BSD seem to be compatible), it seems we could also just incorporate ```pyexcel``` + plugins into ```pandas``` instead.

An alternative strategy could be to see if the ```pyexcel``` maintainer is open for maintaining/supporting a version that includes all the plugins, but one that fails graciously when the required readers are not installed (as is done now in ```pandas```).

Either way, I don't have any strong feelings regarding the current approach or something different like ```pyexcel```. After all, the code that is specific to a certain reader is limited, and writing it specifically with ```pandas``` in mind might be more suitable for certain edge cases. 

Difficult to say, but migrating to ```pyexcel``` might be more work compared to implementing the missing features into ```pandas```.

**Comment by davidovitch** [comment]
I have rebased my work again, and squashed a lot of intermediate steps into one the ```ENH``` commit. I hope this will help the review process.

File: pandas/io/excel.py [hunk 3]
```diff
@@ -66,27 +69,164 @@ def get_writer(engine_name):
...
+        if len(args) > 0 and isinstance(args[0], self.io_class):
+            return args[0]
+        return self._open_workbook(*args, **kwargs)
+
+    def is_ext(self, path):
+        """Verify if the path's extension is supported by the reader. If the
+        reader is not available, return False
+        """
+        ext = path.split('.')[-1]
+        if isinstance(self.io_class, type(None)):
```
Comment (shoyer) [inline pandas/io/excel.py]: I think this should just be `self.is_class is None`

File: pandas/io/excel.py [hunk 4]
```diff
@@ -66,27 +69,164 @@ def get_writer(engine_name):
...
+    # fil types, see PR #9070
+
+    def __init__(self, engine, extensions):
+        """Set the engine name, and extension. If the engine is not installed,
+        io_class and _open_workbook are type(None) and None respectively,
+        but the ImportError is ingored at the init stage.
+        """
+        self.engine = engine
+        self.extensions = extensions
+        self.io_class = type(None)
```
Comment (shoyer) [inline pandas/io/excel.py]: I would just use `None` here...

File: pandas/io/excel.py [hunk 5]
```diff
@@ -66,27 +69,164 @@ def get_writer(engine_name):
...
+        self._open_workbook = None
+        # fail silently if engine is not presented on init, only raise when
+        # attempting to use it when opening a workbook
+        try:
+            self.load_engine()
```
Comment (shoyer) [inline pandas/io/excel.py]: what is the purpose of this messy fallback logic?

File: pandas/io/tests/test_excel.py [hunk 6]
```diff
@@ -58,6 +59,13 @@ def _skip_if_no_xlsxwriter():
         raise nose.SkipTest('xlsxwriter not installed, skipping')
 
 
+def _skip_if_no_ezodf():
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: So instead of this, make a new classes, ``ODSReaderTests/ExcelReaderTests``, and a new base class, ``ReadingTests`` where these both inherit. Then almost all of the existing ``ExcelReaderTests`` can be moved to ``ReadingTests``. Use a class variable to hold the suffixes. Add a new function:

```
# for Excel
def _skip_if(self):
    _skip_if_no_xlrd()     

# for ODS
def _skip_if(self):
    _skip_if_no_ezodf()
```

Then remove all of the ``_skip_if_no......`` and replace with ``self._skip_if()``

File: pandas/io/excel.py (hunk 5 above)
Comment (davidovitch) [inline pandas/io/excel.py]: What I am trying to do here is to be able to create a file reader class that does not fail if its dependencies are not installed. The logic for deriving the excel reader type now iterates over the supported types, regardless of the availability of their dependencies. In that way the file type can still be determined for certain cases, and give the user a more clear error message when a dependency is missing. The switching logic is in ```ExcelFile()```, starting from line nr [313](https://github.com/pydata/pandas/pull/9070/files#diff-0153a977179cfc73a2645cf0b7cd1f7fR313).

File: pandas/io/excel.py (hunk 3 above)
Comment (davidovitch) [inline pandas/io/excel.py]: Yeah, I think you have a good point, using ```None``` is more elegant. Thanks.

**Comment by davidovitch** [comment]
I re-factored the tests so the structure is consistent with how the Excel writer tests are implemented. If I got it right this should be address @jreback comments. Previously, not all tests where executed for all the 3 different Excel formats (because not all files where available in all 3 formats. Now all test spreadsheets are available in the 3 formats (```xls```, ```xlsx```, ```xlsm```)

Although I agree with  @shoyer that using ```None``` instead of ```type(None)``` for ```io_class``` seems more elegant at first sight, ```None``` will fail when using ```isinstance(object, None)```. Since at some places I need to test if a certain object is possibly something else then None, I kind of assume it is nicer to have:
```python
if isinstance(object, a_type_that_can_be_NoneType):
    blah
```
over 
```python
if object is not None and isinstance(object, a_type_that_cannot_be_NoneType):
    blah
```

I will squash and rebase in a few days, but I will retain the re-factoring of the reader tests in a separate commit.

**Comment by jreback** [comment]
using type(None) is not idiomatic at all
pls change this

**Comment by davidovitch** [comment]
Sure.

Is there any way of avoiding something like this instead?
```Python
if self.io_class is None:
    return False
elif isinstance(io, self.io_class):
    return True
else:
    return False
```
I can see how ```type(None)``` violate Python idioms; I was trying to be consistent on the data type of io_class: It can be either ```None```, or a type (```xlrd.Book``` or ```ezodf.document.PackagedDocument```). But maybe that doesn't matter at all :-)

**Comment by jreback** [comment]
return isinstance(io, self.io_class)

**Comment by davidovitch** [comment]
but that will fail if ```self.io_class``` is ```None```... or am I seeing ghosts?

**Comment by jreback** [comment]
nope

**Comment by davidovitch** [comment]
That's weird, what I get is:
```
>>> isinstance(True, True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isinstance() arg 2 must be a type or tuple of types
>>> isinstance(True, type(True))
True
>>> 
```

**Comment by jreback** [comment]
why would io_class be None

**Comment by jreback** [comment]
or just do

return io is None or isinstnace(io, io_class)

**Comment by davidovitch** [comment]
Because if the specific reader is not installed, it can be None. 
Yeah, that's way more compact :-) but the same principle still applies.

File: pandas/io/excel.py (hunk 5 above)
Comment (davidovitch) [inline pandas/io/excel.py]: But maybe this approach is not the most lean, clean and readable (based on your comments at least ;-). What I see as the challenge is that a user could try to read a spreadsheet that is supported by pandas, but not installed. I could try to re-factor this in another way. Instead of having some kind of fall back logic in the base/sub class of the reader, there could be something that registers which reader dependencies are a installed. When knowing which readers are available and which ones are not, I assume this could be made nicer. Do you think it is necessary to revise the current strategy?

File: pandas/io/excel.py [hunk 7]
```diff
@@ -66,27 +69,162 @@ def get_writer(engine_name):
...
+            raise NotImplementedError(msg)
+
+
+class XLRDFile(BaseFile):
+    """File reader class for MS Excel spreadsheets (depends on xlrd)
+    """
+    def __init__(self):
+        # engine, extensions, are defined here, but io_class and open_workbook
+        # are only defined when importing the engine
+        args = ('xlrd', ['xls', 'xlsx', 'xlsm'])
```
Comment (jreback) [inline pandas/io/excel.py]: the extensions should be a class property

File: pandas/io/excel.py [hunk 8]
```diff
@@ -66,27 +69,162 @@ def get_writer(engine_name):
...
+        """Verify if the io type is supported by the reader. If the reader is
+        not installed, return False.
+        """
+        if io is None or self.io_class is None:
+            return False
+        return isinstance(io, self.io_class)
+#        elif self.engine == engine:
+#            return True
+#        elif isinstance(io, compat.string_types):
+#            if self.is_ext(io):
```
Comment (jreback) [inline pandas/io/excel.py]: not sure what the point of this function. You are only instantiating a single ``BaseFile`` instance based on the extension.

File: pandas/io/excel.py [hunk 9]
```diff
@@ -66,27 +69,162 @@ def get_writer(engine_name):
...
+    If the engine is not installed, io_class and _open_workbook are
+    type(None) and None respectively. When attempting to use open_workbook
+    while _open_workbook is None, the relevant ImportError is raised. When
+    calling is_type or is_ext, False is retunred when the reader is not
+    installed.
+    """
+    # added in order to deal with both Open Document Format and MS Excel
+    # fil types, see PR #9070
+
+    def __init__(self, engine, extensions):
```
Comment (jreback) [inline pandas/io/excel.py]: make a class-method ``BaseFile.create`` which returns the appropriate sub-class based on the extension. Then you don't really need much of this class/logic.

File: pandas/io/excel.py [hunk 10]
```diff
@@ -154,44 +292,57 @@ def read_excel(io, sheetname=0, **kwds):
...
-            if _is_url(io):
-                data = _urlopen(io).read()
-                self.book = xlrd.open_workbook(file_contents=data)
-            else:
-                self.book = xlrd.open_workbook(io)
-        elif engine == 'xlrd' and isinstance(io, xlrd.Book):
-            self.book = io
-        elif not isinstance(io, xlrd.Book) and hasattr(io, "read"):
+        self.engine = kwds.pop('engine', None)
+        # Initiate pandas supported file readers
```
Comment (jreback) [inline pandas/io/excel.py]: instead of creating instances here. should be something like:

``self.engine, self.book = BaseFile.create(io, engine=engine)``

you can move almost all of this code to BaseFile

**Comment by davidovitch** [comment]
@jreback @shoyer: thanks again for your comments! Let me know if you prefer a fresh squash/rebase for having another look at this.

File: pandas/io/excel.py [hunk 11]
```diff
@@ -66,27 +69,211 @@ def get_writer(engine_name):
...
+        self.engine = engine
+        self.io_class = None
+        self._open_workbook = None
+        # fail silently if engine is not presented on init, only raise when
+        # attempting to use it when opening a workbook
+        try:
+            self.load_engine()
+        except ImportError:
+            pass
+#        if io is not None:
```
Comment (shoyer) [inline pandas/io/excel.py]: commented out code should be removed

File: pandas/io/tests/test_excel.py [hunk 12]
```diff
@@ -92,110 +95,122 @@ def read_csv(self, *args, **kwds):
...
+    #    installed.
+    # 2. Add a property ext, which is the file extension that your writer
+    #    writes to. (needs to start with '.' so it's a valid path)
+    # 3. Add a property engine_name, which is the name of the writer class.
+    #    For the reader this is not used for anything at the moment.
+
+    def setUp(self):
+        self.check_skip()
+        super(ReadingTestsBase, self).setUp()
+#        self.option_name = 'io.excel.%s.reader' % self.ext.strip('.')
```
Comment (shoyer) [inline pandas/io/tests/test_excel.py]: this commented out code should be removed

File: pandas/io/tests/test_excel.py [hunk 13]
```diff
@@ -92,110 +95,122 @@ def read_csv(self, *args, **kwds):
...
-            df2 = df2.reindex(columns=['A', 'B', 'C'])
-            df3 = xls.parse('Sheet2', skiprows=[1], index_col=0,
-                            parse_dates=True, parse_cols=3)
-            # TODO add index to xls file)
-            tm.assert_frame_equal(df, df2, check_names=False)
-            tm.assert_frame_equal(df3, df2, check_names=False)
+        pth = os.path.join(self.dirpath, 'test1' + self.ext)
+        pref = os.path.join(self.dirpath, 'test1.csv')
+        # the truth is obtained form read_csv with Python engine
+        dfref = self.read_csv(pref, index_col=0, parse_dates=True)
```
Comment (shoyer) [inline pandas/io/tests/test_excel.py]: This block of code is repeated quite a few times in the test... can you make it a method instead?

File: pandas/io/tests/test_excel.py [hunk 14]
```diff
@@ -92,110 +95,122 @@ def read_csv(self, *args, **kwds):
...
+        xls = ExcelFile(pth)
+        df = xls.parse('Sheet1', index_col=0, parse_dates=True,
+                       parse_cols='A:D')
+        df2 = dfref.reindex(columns=['A', 'B', 'C'])
+        df3 = xls.parse('Sheet2', skiprows=[1], index_col=0,
+                        parse_dates=True, parse_cols='A:D')
+        # TODO add index to xls, read xls ignores index name ?
+        tm.assert_frame_equal(df, df2, check_names=False)
+        tm.assert_frame_equal(df3, df2, check_names=False)
+        del df, df2, df3
```
Comment (shoyer) [inline pandas/io/tests/test_excel.py]: why are you explicitly deleted these variables? generally there's no reason to do that in python....

File: pandas/io/excel.py [hunk 15]
```diff
@@ -66,27 +69,211 @@ def get_writer(engine_name):
...
+        ext = path.split('.')[-1]
+        if self.io_class is None:
+            return False
+        return ext in self.extensions
+
+    def load_engine(self):
+        """Load the engine if installed, but raise ImportError otherwise.
+        If engine is not defined, do nothing.
+        """
+        try:
```
Comment (jreback) [inline pandas/io/excel.py]: I don't think you need this double exception here. the ``self._load_engine()`` should simply raise the appropriate exception.

File: pandas/io/excel.py [hunk 16]
```diff
@@ -66,27 +69,211 @@ def get_writer(engine_name):
...
+            Engine used for reading the io
+
+        book : object
+            The spreadsheet book class created by the engine.
+        """
+        self.engine = engine
+        # Initiate pandas supported file readers. When creating new readers,
+        # add them to the supported_readers list
+        xlrd_f = XLRDFile()
+        ezodf_f = EZODFFile()
```
Comment (jreback) [inline pandas/io/excel.py]: why don't you make ``supported_readers`` a cached property which returns this list. This should return readers which don't raise import errors.

File: pandas/io/excel.py [hunk 17]
```diff
@@ -66,27 +69,211 @@ def get_writer(engine_name):
...
+                    return self.engine, self.book
+            # Does the io type match any supported and installed reader types?
+            # Note that io_class is None if engine is not installed
+            elif reader.io_class is not None and isinstance(io, reader.io_class):
+                self.engine = reader.engine
+                self.book = io
+                return self.engine, self.book
+
+        # xlrd has some additional/alternative reading mechanisms:
+        if not isinstance(io, xlrd_f.io_class) and hasattr(io, "read"):
```
Comment (jreback) [inline pandas/io/excel.py]: this can be encapsulated in a method for each reader, e.g. in the xlrd, ezodf sub-classes, e.g. something like
hen the loop becomes

```
for reader in self.supported_readers:
    try:
          return reader.create_engine(io, engine=engine):
    except (NotImplementedError):
         pass

raise ValueError("no supported readers for this input")
```

**Comment by davidovitch** [comment]
@shoyer @jreback and once more thanks for reviewing (hope this PR is not too boring for you guys...). I have tried to address all you comments.

I am not sure I see how to get around @jreback last suggestion: 
> why don't you make supported_readers a cached property which returns this list. This should return readers which don't raise import errors.
> this can be encapsulated in a method for each reader, e.g. in the xlrd, ezodf sub-classes, e.g. something like then the loop becomes
> ```python
for reader in self.supported_readers:
    try:
          return reader.create_engine(io, engine=engine):
    except (NotImplementedError):
         pass
raise ValueError("no supported readers for this input")
```
This approach would require another mechanism to throw and ```ImportError ``` when the reader corresponding the input is supported but not installed. So I am not sure I fully understand how to get the last suggestion implemented properly.

Don't get me wrong, I am getting a lot out of your comments and suggestions, and I don't want to give up just yet (I would love to see pandas read my ods spreadsheets), but I would like to ask your opinion on how this is going as this is my first "significant" PR (I am not educated as a coder). Do you think the PR is going in the right direction? Are the "improvements" I am making getting the PR to an acceptable level? If I am not able to clear the quality bar, I would rather have someone more capable and experienced take this over instead. It would be a shame if am just wasting your time over a PR that is not going anywhere.

**Comment by andresmrm** [comment]
Just to let you know, I'm watching this thread and waiting eagerly for this feature. =)

**Comment by shoyer** [comment]
I'm on my phone now, but if I recall correctly the code here was indeed on the right track. The key thing is that it's well tested -- clean code can come later.

**Comment by jreback** [comment]
@davidovitch I'll have a look tomorrow
but the point is tests are key - and the code should be pretty idiomatic
I have found that it is best to put in pretty tight code on the first pass - yes things can and do get revised afterwards - but that is harder esp if the original authors are not around (which can be frequent) and we already have enough to do w/o auditing all parts of the code

**Comment by davidovitch** [comment]
Thanks, I completely agree that the code should be well tested and idiomatic. Just wanted to make sure my coding-, and your review-efforts are leading up to something. I am not in any particular hurry, and would like to sit this one through until the end :-) (for whatever number of iterations remain).

File: pandas/io/excel.py [hunk 18]
```diff
@@ -68,27 +72,207 @@ def get_writer(engine_name):
...
+class BaseFile(object):
+    """Base class for excel readers
+
+    A file class can be initialized even if the engine is not installed.
+    If the engine is not installed, io_class and _open_workbook are None.
+    When attempting to use open_workbook while _open_workbook is None, the
+    relevant ImportError is raised. When calling is_ext, False is retunred when
+    the reader is not installed.
+    """
+    # added in order to deal with both Open Document Format and MS Excel
```
Comment (jreback) [inline pandas/io/excel.py]: comment not necessary

File: pandas/io/excel.py [hunk 19]
```diff
@@ -68,27 +72,207 @@ def get_writer(engine_name):
...
+    # fil types, see PR #9070
+
+    def __init__(self, engine=None):
+        """Set the engine name, and extension. If the engine is not installed,
+        io_class and _open_workbook are both None.
+        """
+        self.engine = engine
```
Comment (jreback) [inline pandas/io/excel.py]: you don't need to pass engine. Its simply a class attribute.

File: pandas/io/excel.py [hunk 20]
```diff
@@ -68,27 +72,207 @@ def get_writer(engine_name):
...
+        return self.open_workbook(file_contents=data)
+
+
+class EZODFFile(BaseFile):
+    """File reader class for ODF spreadsheets (depends on ezodf)
+    """
+    extensions = ['ods']
+
+    def __init__(self, engine='ezodf'):
+        # engine, extensions, are defined here, but io_class and open_workbook
```
Comment (jreback) [inline pandas/io/excel.py]: so just set the engine as as a class attribute, e.g. here would be ``engine='ezodf'``. you don't need the ``__init__`` method in the sub-class then.

File: pandas/io/excel.py [hunk 21]
```diff
@@ -68,27 +72,207 @@ def get_writer(engine_name):
...
+    # added in order to deal with both Open Document Format and MS Excel
+    # fil types, see PR #9070
+
+    def __init__(self, engine=None):
+        """Set the engine name, and extension. If the engine is not installed,
+        io_class and _open_workbook are both None.
+        """
+        self.engine = engine
+        self.io_class = None
+        self._open_workbook = None
```
Comment (jreback) [inline pandas/io/excel.py]: might has well make this ``.open_workbook``. confusing otherwise.

Comment (jreback) [inline pandas/io/excel.py]: call this ``workbook_factory`` or something

File: pandas/io/excel.py [hunk 22]
```diff
@@ -68,27 +72,207 @@ def get_writer(engine_name):
...
+        # try to load the engine again and raise import error if required
+        if self._open_workbook is None:
+            self.load_engine()
+        # just in case the user passes an already opened workbook of io_class
+        if len(args) > 0 and isinstance(args[0], self.io_class):
+            return args[0]
+        return self._open_workbook(*args, **kwargs)
+
+    def create(self, io, engine=None):
+        """Create the appropriate speadsheet book based on io and optionally
```
Comment (jreback) [inline pandas/io/excel.py]: this is VERY confusing what you are doing. ``.create`` is generating returning engine/book, but you aren't using the class sructure at all.

This should be a class method and simply return an instance of the appropriate sub-class, something like:

``parser = BaseFile.create(io, engine=engine)``


then it has properties for everything you need

``parser.engine, parser.book, parser.parser``, whatever.

File: pandas/io/excel.py [hunk 23]
```diff
@@ -436,18 +602,182 @@ def _parse_cell(cell_contents,cell_typ):
...
             output[asheetname] = parser.read()
-            
+
         if ret_dict:
             return output
         else:
             return output[asheetname]
-        
+
+    def _parse_ods(self, sheetname=0, header=0, skiprows=None, skip_footer=0,
```
Comment (jreback) [inline pandas/io/excel.py]: so this should be a method on ``EZDFile`` called ``.parse``. (and the current ``._parse_excel`` should be on ``XLRDFile`` with the same name.

File: pandas/io/excel.py [hunk 24]
```diff
@@ -156,47 +340,22 @@ def read_excel(io, sheetname=0, **kwds):
...
-        elif engine == 'xlrd' and isinstance(io, xlrd.Book):
-            self.book = io
-        elif not isinstance(io, xlrd.Book) and hasattr(io, "read"):
-            # N.B. xlrd.Book has a read attribute too
-            data = io.read()
-            self.book = xlrd.open_workbook(file_contents=data)
-        else:
-            raise ValueError('Must explicitly set engine if not passing in'
-                             ' buffer or path for io.')
+        self.engine, self.book = BaseFile().create(io, engine=engine)
```
Comment (jreback) [inline pandas/io/excel.py]: you want to save the file object

e.g.

```
self.factory = BaseFile.create(io, engine=engine)
```

``.engine, .book`` are just properties on the factory

**Comment by jorisvandenbossche** [comment]
@davidovitch do you have some time to deal with the latest comments?

Would be nice to get this into 0.17.0!

**Comment by davidovitch** [comment]
Yes, sorry for the late response again...I will try to get this done ASAP. What is the 0.17 time schedule?

**Comment by jreback** [comment]
closing things in about 2 weeks, then release candidate & final in mid sept.

**Comment by davidovitch** [comment]
ok, by mid next week I should have things sorted out.

**Comment by davidovitch** [comment]
I've tried to address the comments again in the last 3 commits (scroll upwards to see the 3th). @jreback, does this address your comments specifically? Note that now the ```ExcelFile``` attributes ```engine``` and ```book``` are moved to ```factory.engine``` and ```factory.book```. If anyone used these attributes, this change will brake those codes. There is room for improvement, but the only way out I see is by refactoring the whole thing and follow a similar approach as the writers.

I am working on creating an additional test file to test the datetime and timedelta conversions for the EzODF reader. I'll push them before the end of the week.

Let me know if you prefer a rebase based on upstream before reviewing again.

**Comment by jreback** [comment]
You are still doing all kinds of if/then all over the place, and the logic is still quite convoluted. Pls try to simplify according to all of my previous comments.

**Comment by davidovitch** [comment]
It is not for the lack of trying (you will probably not belief the amount of time I already spent on this PR), or for ignoring you comments, it is that I am failing to properly translate the comments into clean, Pythonic code that clears the quality checks of the review procedure. I agree with your assessment, but I can only conclude that I fail to boil it down into simple and clean logic.

I have some more time today and tomorrow to work on this PR, lets see where we stand after that.

**Comment by jreback** [comment]
@davidovitch no I believe you are trying hard! let's squash down to a single commit and rebase. I won't have time to play with this for a bit though.

**Comment by davidovitch** [comment]
@jreback thanks for your patience. Too bad I couldn't make it for the 0.17 release, but at least I tried and learned something in the process :-)

Before squashing everything into a single commit, what I think makes sense to do is the following:
* I create a new PR that contains the refactoring of the writer tests in ```test_excel.py```. This PR would also contain the ```ods``` test files, and the renamed MS Excel files (```xls```, ```xlsx```, ```xlsm```). I've renamed them so it is more consistent, and corresponds to the ```ods``` files. This refactoring was necessary to ease the processes of supporting multiple reader mechanisms. Regardless of the outcome of the actual implementation of the new ods reader, I think this work got to an acceptable level, and can be merged independent from the actual new ODS reader. Not that is such a significant contribution, but it would otherwise get kind of lost (or create a more laborious merge later if other changes have appeared in the mean time). I would suggest to take this one up into the 0.17 release.
* I close this PR and create a new PR that only contains the changes to the reader in a single commit. In that way this messy PR and its history is kept for "the record". In the description of the new PR I summarize with what I struggled, and some of the key comments of the reviewers, and refer to the old PR.
* I think it would be great if someone could step in to fix this faster and better then I can, but until that happens, I hope it is ok if I try to find away to figure out how to do it more lean, clean and Pythonic (but I will not bother the reviewers again until I really think the results have improved significantly).

If you think this it too much whatever, please say so and I will just squash into one commit and forgot about the whole thing, no hard feelings :-)

@andresmrm, if you really want this badly, you can always give this fork a spin. The current implementation actually works, and passes all the tests (if you have a lot of ODS files hanging around you can test it more extensively and report back any bugs in this PR ;-) ). However, the implementation is not on par with the code quality everyone expects from a library as Pandas.

**Comment by timmie** [comment]
@davidovitch 
Off topic: I understand fully your situation. I was in a similar case some time ago. Occasional users who are not trained software engineers have difficulties to provide contributions to the level required at pandas.
Ironically, it was with the excel reader. My code worked, passed the test, but style and technical issues remained leaving the work stalled due to time and understanding:
https://github.com/pydata/pandas/issues/4340
https://github.com/pydata/pandas/pull/4404
https://github.com/pydata/pandas/pull/4631

Good luck and perseverance.

**Comment by jreback** [comment]
@davidovitch ok on a refactored test pr (though don't include the ods files; just the new naming scheme)

**Comment by davidovitch** [comment]
@jreback, the refactored tests (excluding ODS files and tests as requested) are in PR #10964.
I've moved the EzODF powered ods reader into the branch [ezodf_reader](https://github.com/davidovitch/pandas/tree/ezodf_reader) of my pandas fork.

**Comment by hnykda** [comment]
I am lost in all related PRs. What is the status of this feature (why is it closed)? I am willing to contribute somehow, but really don't know where to start.

**Comment by jreback** [comment]
well, you can start with this PR and create a new one that addresses the comments.

**Comment by jreback** [comment]
contributing guide is [here](http://pandas.pydata.org/pandas-docs/stable/contributing.html)

**Comment by davidovitch** [comment]
@hnykda I would start with rebasing the fork again on top of the current pandas master. I've been planning to do this for a while, but I haven't got around it. Let me know if you manage/plan to do it so I don't need to that again ;-) Please feel free to dig in or ask any questions you have regarding the current implementation.

**Comment by hnykda** [comment]
Thanks. I unfortunately don't see that in the near future, probably after the new year.

**Comment by davidovitch** [comment]
@hnykda I've rebased the [ezodf_reader](https://github.com/davidovitch/pandas/tree/ezodf_reader) branch in my pandas fork with the pandas master in case you like to have a look at it. I might try to improve the current implementation in the near future.

**Comment by TrigonaMinima** [comment]
@davidovitch any plans of working on this or can I take it up?

**Comment by detrout** [comment]
For what its worth I wrote a ODS reader that produces a pandas DataFrame with tests for the various ways my spreadsheets became hard to parse.

I tried to stay close to the pandas _XlrdReader class 

https://github.com/detrout/pandasodf

**Comment by jreback** [comment]
@detrout that looks pretty good - seems way enough to adapt if you’d like to do a PR

**Comment by detrout** [comment]
@jreback I hadn't done a PR because I hadn't bothered to implement a writer. Is it worth going ahead and submitting just the reader class?

**Comment by WillAyd** [comment]
@detrout that is totally fine

**Comment by TrigonaMinima** [comment]
@detrout, that's really clean code. I was also thinking of going the ```odfpy``` route myself. You should submit a PR since you have already finished some major chunk of it. If you won't be able to then please let me know.

**Comment by detrout** [comment]
@TrigonaMinima Ok I'll try to put together over the weekend.

**Comment by detrout** [comment]
Quick question.

odfpy uses ISO 8601 durations and one of my ods test cases has a timeduration, of PT03H45M00S. I had my own handrolled parser, because I wrote this a while ago. But pandas now has a iso8601 duration parser.

Unfortunately it assumes that there will always be a day specifier and fails with PT03H45M00S, but does work with P0DT03H45M00S.  So I filed https://github.com/pandas-dev/pandas/issues/25422

Below is the quick workaround I came up with.

```
def pandas_isoduration_normalization(duration):
    """Make libreoffice durations compatible with current pandas Timedelta parser

    The current parser can't handle durations that lack a time component 
    to the left of the of the T. 
    For example PT3H45M0S.
    """
    if duration.startswith('PT'):
        duration = 'P0DT' + duration[2:]
    return pandas.Timedelta(duration)
```

Is that acceptable for the moment?

**Comment by jreback** [comment]
yes that’s ok but if u can do as a separate PR first 
alternatively u can do it for this one and can separate later

**Comment by detrout** [comment]
My rough draft PR with an alternate implementation is here https://github.com/pandas-dev/pandas/pull/25427
